rust/noosphere/src/ffi/petname.rs

```diff
@@ -1,47 +1,173 @@
-/// the ID of the sphere, but in order to read the sphere you must resolve the
-/// DID to a CID, which tells you the version of the sphere to read.
+/// Get the sphere identity - a DID - assigned to a given petname of a sphere.
+/// Note that this call will produce an error if no address has been assigned to
+/// the name.
 pub fn ns_sphere_petname_get(
-    noosphere: &NsNoosphereContext,
-    sphere_identity: char_p::Ref<'_>,
+    noosphere: &NsNoosphere,
+    sphere: &NsSphere,
     petname: char_p::Ref<'_>,
     error_out: Option<Out<'_, repr_c::Box<NsError>>>,
 ) -> Option<char_p::Box> {
-    todo!();
+    error_out.try_or_initialize(|| {
+        noosphere.async_runtime().block_on(async {
+            sphere
+                .inner()
+                .get_petname(petname.to_str())
+                .await?
+                .ok_or_else(|| anyhow!("No petname '{}' has been set", petname.to_str()))?
+                .to_string()
+                .try_into()
+                .map_err(|error: InvalidNulTerminator<String>| anyhow!(error).into())
+        })
+    })
 }
 
 #[ffi_export]
-/// Assign a DID to a petname. This will overwrite a petname entry if one already exists
-/// with the given name (and reset the resolved CID, if any).
+/// Assign a petname to a sphere identity (a DID). This will overwrite the
+/// petname with the new sphere identity if it has been set previously. When a
+/// new address is assigned to a petname, the petname will be reset to an
+/// unresolved state. Make sure to invoke sync after setting new petnames, so
+/// that new sphere identities can be resolved by the gateway (if one is
+/// configured).
 pub fn ns_sphere_petname_set(
-    noosphere: &NsNoosphereContext,
-    sphere_identity: char_p::Ref<'_>,
+    noosphere: &NsNoosphere,
+    sphere: &mut NsSphere,
     petname: char_p::Ref<'_>,
-    did: char_p::Ref<'_>,
+    did: Option<char_p::Ref<'_>>,
     error_out: Option<Out<'_, repr_c::Box<NsError>>>,
 ) {
-    todo!();
+    error_out.try_or_initialize(|| {
+        noosphere.async_runtime().block_on(async {
+            sphere
+                .inner_mut()
+                .set_petname(petname.to_str(), did.map(|did| did.to_str().into()))
+                .await?;
+
+            Ok(())
+        })
+    });
 }
 
 #[ffi_export]
-/// Resolve a configured petname to a sphere version (a CID), via the DID that
-/// has been assigned to it. The returned value is a UTF-8, base64-encoded CIDv1
-/// string. If no DID has been assigned to the given petname, no value will be
-/// resolved.
+/// Resolve a configured petname, using the sphere idenitty - a DID - to
+/// determine a link - a CID - that is associated with it. The returned link is
+/// a UTF-8, base64-encoded CIDv1 string that may be used to resolve data from
+/// the IPFS content space. Note that this call will produce an error if no
+/// address has been assigned to the given petname.
```

Comment:
Unclear what scenario produces an error, as "address" is unclear -- "address" could be "identity" (referring to a bound DID) or "link" (referring to a NS-resolved CID), and IIUC, two different scenarios could result in this. Worth renaming "address" to either "identity" or "link" (IMO, I think most of the comments are pretty consistent with it), and including all scenarios that produce error (which may be if either "link" or "identity" are absent).

Reply:
An address can never be link, this is just a typo where the word address should be identity.